WeakAuras/WeakAuras.lua

```diff
@@ -1506,64 +1505,94 @@ Broker_WeakAuras = LDB:NewDataObject("WeakAuras", {
-  end
-  coroutine.yield();
+function WeakAuras.IsLoginFinished()
+  return loginFinished
+end
 
-  WeakAuras.AddMany(toAdd);
-  coroutine.yield();
-  WeakAuras.AddManyFromAddons(from_files);
-  WeakAuras.RegisterDisplay = WeakAuras.AddFromAddon;
-  coroutine.yield();
-  WeakAuras.ResolveCollisions(function() registeredFromAddons = true; end);
-  coroutine.yield();
+function WeakAuras.LoginMessage()
+  return loginMessage
+end
 
-  for _, triggerSystem in pairs(triggerSystems) do
-    if (triggerSystem.AllAdded) then
-      triggerSystem.AllAdded();
-      coroutine.yield();
+local isDatabaseUpgrade = false
+function WeakAuras.IsDatabaseUpgrade()
+  return isDatabaseUpgrade
+end
+
+function WeakAuras.Login(initialTime, takeNewSnapshots)
+  local loginThread = coroutine.create(function()
+    WeakAuras.Pause();
+    local toAdd = {};
+    for id, data in pairs(db.displays) do
+      if(id ~= data.id) then
+        print("|cFF8800FFWeakAuras|r detected a corrupt entry in WeakAuras saved displays - '"..tostring(id).."' vs '"..tostring(data.id).."'" );
+        data.id = id;
+      end
+      tinsert(toAdd, data);
     end
-  end
+    coroutine.yield();
 
-  -- check in case of a disconnect during an encounter.
-  if (db.CurrentEncounter) then
-    WeakAuras.CheckForPreviousEncounter()
-  end
-  coroutine.yield();
-  WeakAuras.RegisterLoadEvents();
-  WeakAuras.Resume();
-  coroutine.yield();
+    isDatabaseUpgrade = takeNewSnapshots;
+    WeakAuras.AddMany(toAdd);
+    coroutine.yield();
+    WeakAuras.AddManyFromAddons(from_files);
+    WeakAuras.RegisterDisplay = WeakAuras.AddFromAddon;
+    coroutine.yield();
+    WeakAuras.ResolveCollisions(function() registeredFromAddons = true; end);
+    coroutine.yield();
+    isDatabaseUpgrade = false;
 
-  local nextCallback = loginQueue[1];
-  while nextCallback do
-    tremove(loginQueue, 1);
-    if type(nextCallback) == 'table' then
-      nextCallback[1](unpack(nextCallback[2]))
-    else
-      nextCallback()
+    for _, triggerSystem in pairs(triggerSystems) do
+      if (triggerSystem.AllAdded) then
+        triggerSystem.AllAdded();
+        coroutine.yield();
+      end
+    end
+
+    -- check in case of a disconnect during an encounter.
+    if (db.CurrentEncounter) then
+      WeakAuras.CheckForPreviousEncounter()
     end
     coroutine.yield();
-    nextCallback = loginQueue[1];
-  end
+    WeakAuras.RegisterLoadEvents();
+    WeakAuras.Resume();
+    coroutine.yield();
 
-  loginFinished = true
-  WeakAuras.ResumeAllDynamicGroups();
-end)
+    local nextCallback = loginQueue[1];
+    while nextCallback do
+      tremove(loginQueue, 1);
+      if type(nextCallback) == 'table' then
+        nextCallback[1](unpack(nextCallback[2]))
+      else
+        nextCallback()
+      end
+      coroutine.yield();
+      nextCallback = loginQueue[1];
+    end
 
-function WeakAuras.IsLoginFinished()
-  return loginFinished
-end
+    loginFinished = true
+    WeakAuras.ResumeAllDynamicGroups();
```

Comment:
This is problematic if Login() is called twice, see later comment

Reply:
Actually it might be okay, seems ResumeAllDynamicGroups doesn't decrease below 0, so it shouldn't matter.

Although I question that ResumeAllDynamicGroups, I thought that dynamic groups were created suspended, and that's why we do that. But I can't find any code that suspends newly created dynamic groups, not code that initially suspend it.